qiskit/circuit/instruction.py

```diff
@@ -136,34 +137,36 @@ def params(self, parameters):
         self._params = []
         for single_param in parameters:
             # example: u2(pi/2, sin(pi/4))
-            if isinstance(single_param, (ParameterExpression, sympy.Basic)):
+            if isinstance(single_param, (ParameterExpression)):
                 self._params.append(single_param)
             # example: OpenQASM parsed instruction
             elif isinstance(single_param, node.Node):
-                self._params.append(single_param.sym())
+                self._params.append(single_param.num())
             # example: u3(0.1, 0.2, 0.3)
             elif isinstance(single_param, (int, float)):
-                self._params.append(sympy.Number(single_param))
+                self._params.append(single_param)
             # example: Initialize([complex(0,1), complex(0,0)])
             elif isinstance(single_param, complex):
-                self._params.append(single_param.real +
-                                    single_param.imag * sympy.I)
+                self._params.append(single_param)
             # example: snapshot('label')
             elif isinstance(single_param, str):
-                self._params.append(sympy.Symbol(single_param))
+                self._params.append(float(single_param))
```

Comment:
It looks like the case here allowed for providing string parameters, that would end up in the circuit as `sympy.Symbol`s. e.g.
```
>>> qc = qk.QuantumCircuit(1)
>>> qc.rx('a', 0)
<qiskit.circuit.instructionset.InstructionSet object at 0x131724278>
>>> print(qc)
        ┌───────┐
q_0: |0>┤ Rx(a) ├
        └───────┘
>>> qc.data[0][0]
<qiskit.extensions.standard.rx.RXGate object at 0x10bec90f0>
>>> qc.data[0][0].params[0]
a
>>> type(qc.data[0][0].params[0])
<class 'sympy.core.symbol.Symbol'>
```

Arguably, this is a case where having sympy in the circuit makes the most sense (though now, `circuit.Parameter`s are the better supported option.) We should deprecate this behavior (same for `sympy.Matrix` and `sympy.Expr`).

Reply:
I've update this to create a new `circuit.Parameter` I think this should retain the same behavior as what was there before, but instead use our native symbolic parameter class instead of sympy.

Reply:
I agree that we don't want symbols here long term for all the reasons stated, but I think the right path forward is to make the sympy import as lazy as possible, and deprecate sympy symbols, matrix, and expr as circuit parameters (even if it means we have to wait a bit for the import time improvement, though it should be okay since this is all in the `Instruction.param` setter.)

Generating `circuit.Parameter`s implicitly is something we intentionally waited on, because the rest of the `Parameter` API depends on the user having a reference to the `Parameter` instance (this was our solution to detecting and handling Parameter name conflicts.)

Reply:
ok, I added back support for both sympy.Matrix and sympy.Basic in deprecated form here: https://github.com/Qiskit/qiskit-terra/pull/3156/commits/3e04b847ab3ed25d7fceeeab7429b255458b39c6